lib/card_search_query_builder.rb

```diff
@@ -1,263 +1,52 @@
-                    @parse_error = 'Invalid array operator "%s"' % match_type
-                    return
-                  end
-                  if value.match?(/\A(\w+)-(\d+)\Z/i)
-                    value.gsub!('-', '=')
-                  end
-                  constraints << '%s (? = ANY(%s))' % [operator, @@term_to_field_map[keyword]]
-                  where << value
-                elsif @@boolean_keywords.include?(keyword)
-                    if !['true', 'false', 't', 'f', '1', '0'].include?(value)
-                        @parse_error = 'Invalid value "%s" for boolean field "%s"' % [value, keyword]
-                        return
-                    end
-                    operator = ''
-                    if @@boolean_operators.include?(match_type)
-                        operator = @@boolean_operators[match_type]
-                    else
-                        @parse_error = 'Invalid boolean operator "%s"' % match_type
-                        return
-                    end
-                    constraints << '%s %s ?' % [@@term_to_field_map[keyword], operator]
-                    where << value
-                elsif @@numeric_keywords.include?(keyword)
-                    if !value.match?(/\A(\d+|x)\Z/i)
-                        @parse_error = 'Invalid value "%s" for integer field "%s"' % [value, keyword]
-                        return
-                    end
-                    operator = ''
-                    if @@numeric_operators.include?(match_type)
-                        operator = @@numeric_operators[match_type]
-                    else
-                        @parse_error = 'Invalid numeric operator "%s"' % match_type
-                        return
-                    end
-                    constraints << '%s %s ?' % [@@term_to_field_map[keyword], operator]
-                    where << (value.downcase == 'x' ? -1 : value)
-                else
-                    # String fields only support : and !, resolving to to {,NOT} LIKE %value%.
-                    # TODO(plural): consider ~ for regex matches.
-                    operator = ''
-                    if @@string_operators.include?(match_type)
-                        operator = @@string_operators[match_type]
-                    else
-                        @parse_error = 'Invalid string operator "%s"' % match_type
-                        return
-                    end
-                    constraints << 'lower(%s) %s ?' % [@@term_to_field_map[keyword], operator]
-                    where << '%%%s%%' % value
-                end
-                if @@term_to_left_join_map.include?(keyword)
-                    @left_joins << @@term_to_left_join_map[keyword]
-                end
-             end
+  # TODO(plural): figure out how to do name matches that are LIKEs over elements of an array.
+  # format should implicitly use the currently active card pool and restriction lists unless another is specified.
+  @fields = [
```

Comment:
i had to do a bit of jumping back and forth to track down how the keywords are defined now and have a concern that it will be easy to accidentally clobber something here since the order is important with partial keyword matches.

The comment from the parser about partial matches probably bears repeating here and in the printing version.

Additionally, it may be good to have the keyword be the first element in this struct, not the type of field.

Reply:
I originally had the keyword first, but I also wanted to collect keywords with the same meaning (e.g. `cost` is equivalent to `o`) and putting that array at the start of the field definition makes the formatting way harder to read. As it is now you can see how fields are grouped by type.